File: main_16.py
def parse_rules(rule):
  label, rules = rule.split(':')
  a, b = rules.strip().split(' or ')

  def split_range(rng):
    l, r = rng.split('-')
    return int(l), int(r)

  return (label, [split_range(a), split_range(b)])

def ticket_error_rate(input):
  rules = []
  tickets = []

  sections = 0
  for line in input:
    line = line.strip()
    if line == '':
      continue

    if line.startswith('your ticket:') or line.startswith('nearby tickets:'):
      sections += 1
      continue

    if sections == 0:
      _, rule = parse_rules(line)
      rules.extend(rule)
    if sections == 2:
      tickets.extend(map(int, line.split(',')))

  valid_tickets = {}
  for rule in rules:
    for i in range(rule[0], rule[1]+1):
      valid_tickets[i] = True
  
  total = 0
  for ticket in tickets:
    if ticket not in valid_tickets:
      total += int(ticket)
  return total


def ticket_departure(input):
  rules = {}
  rule_labels = set()
  your_ticket = []
  tickets = []
  valid_tickets = {}
  all_tickets = {}

  sections = 0
  for line in input:
    line = line.strip()
    if line == '':
      continue

    if line.startswith('your ticket:') or line.startswith('nearby tickets:'):
      sections += 1
      continue

    if sections == 0:
      label, rule = parse_rules(line)
      if label not in rule:
        rules[label] = []
        rule_labels.add(label)
      rules[label].extend(rule)

    if sections == 1:
      your_ticket = list(map(int, line.split(',')))
      tickets.append(your_ticket)

    if sections == 2:
      tickets.append(list(map(int, line.split(','))))

  orders = {}
  for i, _ in enumerate(tickets[0]):
    orders[i] = set(list(rule_labels))


  for label in rules:
    rngs = rules[label]
    valid_tickets[label] = {}
    for rng in rngs:
      lo, hi = rng
      for i in range(lo, hi+1):
        valid_tickets[label][i] = True
        all_tickets[i] = True

  for ticket in tickets:
    for i, no in enumerate(ticket):
      if no not in all_tickets:
        continue
      for label in rules:
        if no not in valid_tickets[label]:
          orders[i].remove(label)

  output = {}
  while bool(orders):
    keys = orders.keys()
    for k in keys:
      if len(orders[k]) == 1:
        key = k
    if key is not None:
      output[key] = list(orders[key])[0]
    for o in orders:
      if key != o:
        orders[o].remove(output[key])
    del(orders[key])

  total = None
  for key in output:
    if output[key].startswith('departure'):
      if total is None:
        total = your_ticket[key]
      else:
        total *= your_ticket[key]
  return total

def part_one():
  with open('input_16.txt') as f:
    print(ticket_error_rate(f))


def part_two():
  with open('input_16.txt') as f:
    print(ticket_departure(f))

def main():
  part_one()
  part_two()

if __name__ == '__main__':
  main()

File: main_16.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type Range struct {
	lo int
	hi int
}

func NewRange(lo, hi int) Range {
	return Range{
		lo: lo,
		hi: hi,
	}
}

type Rule struct {
	name   string
	ranges []Range
}

func NewRule(name string, ranges []Range) Rule {
	return Rule{
		name:   name,
		ranges: ranges,
	}
}

func (r Rule) Or(n int) bool {
	for _, rng := range r.ranges {
		if n >= rng.lo && n <= rng.hi {
			return true
		}
	}
	return false
}

func main() {
	var sections int
	var rules []Rule
	var yourTicket []int
	var tickets [][]int
	validTickets := make(map[int]int)
	var columns []int

	parseRange := func(s string) Range {
		ranges := strings.Split(s, "-")
		lo, hi := ranges[0], ranges[1]
		l, err := strconv.Atoi(lo)
		if err != nil {
			log.Fatal(err)
		}
		h, err := strconv.Atoi(hi)
		if err != nil {
			log.Fatal(err)
		}
		return NewRange(l, h)
	}

	parseTickets := func(s []string) []int {
		tickets := make([]int, len(s))
		for i, ticket := range s {
			t, err := strconv.Atoi(ticket)
			if err != nil {
				log.Fatal(err)
			}
			tickets[i] = t
		}
		return tickets
	}

	f, err := os.Open("./input_16.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		txt := scanner.Text()
		if txt == "" {
			continue
		}
		if strings.HasPrefix(txt, "your ticket:") || strings.HasPrefix(txt, "nearby tickets:") {
			sections++
			continue
		}

		switch sections {
		case 0:
			params := strings.Split(txt, ": ")
			name, rawRules := params[0], params[1]
			ranges := make([]Range, 2)
			for i, r := range strings.Split(rawRules, " or ") {
				ranges[i] = parseRange(r)
			}
			rules = append(rules, NewRule(name, ranges))
		case 1:
			yourTicket = parseTickets(strings.Split(txt, ","))
			tickets = append(tickets, yourTicket)
		case 2:
			tickets = append(tickets, parseTickets(strings.Split(txt, ",")))
		default:
		}
	}

	for i, rule := range rules {
		for _, rng := range rule.ranges {
			for j := rng.lo; j <= rng.hi; j++ {
				validTickets[j] |= (1 << i)
			}
		}
	}

	var errorRate int
	for _, ticket := range tickets {
		for _, no := range ticket {
			if _, ok := validTickets[no]; !ok {
				errorRate += no
			}
		}
	}
	fmt.Println("part 1: error rate is", errorRate)

	columns = make([]int, len(rules))

	// For each column set all rules to be valid.
	var allRules int
	for j := range rules {
		allRules |= (1 << j)
	}
	for i := range columns {
		columns[i] = allRules
	}

	for _, ticket := range tickets {
		var valid bool
		for _, no := range ticket {
			_, valid = validTickets[no]
			if valid {
				break
			}
		}
		if !valid {
			continue
		}

		// For each column, check which rules are not valid.
		for i, no := range ticket {
			bit, ok := validTickets[no]
			if !ok {
				continue
			}
			for j := range rules {
				b := 1 << j
				if bit&b != b {
					columns[i] &^= b
				}
			}
		}
	}

	columnMapping := make(map[int]string)
	for len(columnMapping) != len(rules) {
		for i, col := range columns {
			if _, ok := columnMapping[i]; ok {
				continue
			}
			var potentialColumns []Rule
			var k int
			for j := range rules {
				b := 1 << j
				if col&b == b {
					k = j
					potentialColumns = append(potentialColumns, rules[j])
					if len(potentialColumns) > 1 {
						break
					}
				}
			}
			if len(potentialColumns) == 1 {
				columnMapping[i] = potentialColumns[0].name
				for c := range columns {
					columns[c] &^= (1 << k)
				}
			}
		}
	}
	var result int
	for k, v := range columnMapping {
		if strings.HasPrefix(v, "departure") {
			if result == 0 {
				result = yourTicket[k]
			} else {
				result *= yourTicket[k]
			}
		}
	}
	fmt.Println("part 2:", result)
}
